### PR TITLE
Xeno datums

### PR DESCRIPTION
**Pull request by spookydonut**
Removes a lot of mob vars from xenos in favour of datums and removes the mess that is XenoUpgrade.dm

Requires testing.

### CODE REVIEW

File: code/modules/mob/mob_helpers.dm [hunk 1]
```diff
@@ -162,11 +162,16 @@ proc/isXenoDefender(A)
...
 
-proc/isXenoLarva(A) //Xenomorph Larva Hud Test APOPHIS 22MAY2015
+proc/isXenoLarva(A) 
 	if(istype(A, /mob/living/carbon/Xenomorph/Larva))
 		return 1
 	return 0
 
+proc/isXenoLarvaStrict(A)
+	if(istype(A, /mob/living/carbon/Xenomorph/Larva) && !istype(A, /mob/living/carbon/Xenomorph/Larva/predalien))
+		return 1
```
Comment (Rohesie) [inline code/modules/mob/mob_helpers.dm]:
```suggestion
		return TRUE
```

File: code/modules/mob/mob_helpers.dm [hunk 2]
```diff
@@ -162,11 +162,16 @@ proc/isXenoDefender(A)
...
+	return 0
```
Comment (Rohesie) [inline code/modules/mob/mob_helpers.dm]:
```suggestion
	return FALSE
```

File: code/modules/mob/mob_helpers.dm [hunk 3]
```diff
@@ -202,6 +207,11 @@ proc/isXenoWarrior(A)
 		return 1
 	return 0
 
+proc/isXenoPredalien(A)
+	if(istype(A, /mob/living/carbon/Xenomorph/Predalien))
+		return 1
```
Comment (Rohesie) [inline code/modules/mob/mob_helpers.dm]:
```suggestion
		return TRUE
```

File: code/modules/mob/mob_helpers.dm [hunk 4]
```diff
@@ -202,6 +207,11 @@ proc/isXenoWarrior(A)
...
+	return 0
```
Comment (Rohesie) [inline code/modules/mob/mob_helpers.dm]:
```suggestion
	return FALSE
```

File: code/modules/mob/living/carbon/xenomorph/Abilities.dm [hunk 5]
```diff
@@ -993,146 +993,113 @@
...
 		return
-	if(X.observed_xeno)
-		var/mob/living/carbon/Xenomorph/T = X.observed_xeno
-		if(!X.check_plasma(600)) return
+	if(!X.observed_xeno)
+		to_chat(X, "<span class='warning'>You must overwatch the xeno you want to de-evolve.</span>")
+		return
+	
+	var/mob/living/carbon/Xenomorph/T = X.observed_xeno
+	if(!X.check_plasma(600)) return
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/Abilities.dm]: Indentation fetish.

File: code/modules/mob/living/carbon/xenomorph/Abilities.dm [hunk 6]
```diff
@@ -993,146 +993,113 @@
...
+		return
 
-		if(!X.check_state() || !X.check_plasma(600) || X.observed_xeno != T)
-			return
+	if(!X.check_state() || !X.check_plasma(600) || X.observed_xeno != T)
+		return
 
-		if(T.is_ventcrawling)
-			return
+	if(T.is_ventcrawling)
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/Abilities.dm]: Inform the queen of these returns, on why she can't devolve the xeno?

**Comment by Qustinnus** [comment]
a lot of the books in your datum could go under one bit flag so we can easily add new traits if we want to in the future

File: code/modules/mob/living/carbon/xenomorph/Abilities.dm (hunk 6 above)
Comment (spookydonut) [reply code/modules/mob/living/carbon/xenomorph/Abilities.dm]: outside the scope tbh

**Comment by spookydonut** [comment]
> 
> 
> a lot of the books in your datum could go under one bit flag so we can easily add new traits if we want to in the future

Good idea

File: code/modules/mob/living/carbon/xenomorph/Abilities.dm [hunk 7]
```diff
@@ -910,23 +910,26 @@
...
+		return
+	if(target.stat == DEAD)
+		return
+	if(target.health >= target.maxHealth)
+		to_chat(X, "<span class='warning'>[target] is at full health.</span>")
+		return
+	if(X.check_plasma(600))
+		X.use_plasma(600)
+		target.adjustBruteLoss(-50)
+		X.queen_ability_cooldown = world.time + 150 //15 seconds
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/Abilities.dm]:
```suggestion
		X.queen_ability_cooldown = world.time + 15 SECONDS
```

File: code/modules/mob/living/carbon/xenomorph/Abilities.dm [hunk 8]
```diff
@@ -940,21 +943,23 @@
...
+	if(!(target.xeno_caste.caste_flags & CASTE_CAN_BE_GIVEN_PLASMA))
+		to_chat(X, "<span class='warning'>You can't give that caste plasma.</span>")
+		return
+	if(target.plasma_stored >= target.xeno_caste.plasma_max)
+		to_chat(X, "<span class='warning'>[target] is at full plasma.</span>")
+		return
+	if(X.check_plasma(600))
+		X.use_plasma(600)
+		target.gain_plasma(100)
+		X.queen_ability_cooldown = world.time + 150 //15 seconds
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/Abilities.dm]:
```suggestion
		X.queen_ability_cooldown = world.time + 15 SECONDS
```

**Pull request merged**